docs/blueprints/004-research-core-persistent-state.md

```diff
@@ -0,0 +1,110 @@
+Recommended whenever quick access times are needed for large quantities of data.
+
+#### Advantages:
+- Predictable performance: Simple queries like get, put and delete in most cases allows the system’s performance to be a lot more predictable
+- Not based on schema and so you are bound by the model, also changes to the database can be made while in operation
+- Does not require expensive hardware
+- Better performance when compared to relational databases
+- Easy to scale by adding an extra nodes
+
+#### Disadvantages:
+- Simple queries can be complex due to the lack of indexes and scanning capabilities
+- Migrating from one product to another product is difficult in as the API used to access data is store specific
+- Updates requires updating the whole value
+
+#### Example:
+
+##### [BadgerDB](https://dgraph.io/docs/badger/)
+BadgerDB is an embeddable persistent key-value store written in Go. It uses a log-structured merge (LSM) tree based implementation, which has an advantage in throughput when compared to B+ tree since background (disk) writes in LSM maintain a sequential access pattern. Badger supports ACID transactions and Multi-Version Concurrency Control with Snapshot Isolation, it acquires locks on directories when accessing data, so multiple processes cannot open the same database at the same time. If the transaction is read-write, the transaction checks if there is a conflict and return an error if there is one. Regarding queries, it supports Get, Set, Delete, and Iterate functions. Multiple updates can be batched-up into a single transaction, allowing the user to do a lot of writes at a time.
+
+A simple implementation for fuseml-core workflow store using BadgerDB is available at: https://github.com/fuseml/fuseml-core/pull/88, Note that this implementation does not uses BadgerDB directly, instead it uses [BadgerHold](https://github.com/timshannon/badgerhold) which provides a simplified interface for for storing/querying Go types stored on BadgerDB.
+
+Takeaways:
+- Very simple to use as it does not rely on another service. This also has the obvious disadvantage of not working in case of running multiple fuseml-core in parallel, for example in an auto-scale scenario.
+- Fit almost seamlessly the data structure that is already in use by the in-memory store currently implemented in fuseml-core
+- In the case of the workflow store, where most of the objects stored are referred by pointers, the queries does not work as it tries to compare the pointers. But this is mostly a limitation of BadgerHold instead of BadgerDB
```

Comment:
Wait, what? But the actual structs are stored in the store too, not _just_ pointers, am I right?

Reply:
yes, storing the pointers make no sense.